#  Goal of the project
The goal of the project is to generate a value proposition for a new product for book lovers by reviewing information on books, publishers, authors, and customer ratings and reviews, of books from one of the services competing in this market.


# Import libreries and download the data

In [1]:
# import libraries
import pandas as pd
from sqlalchemy import create_engine


In [2]:
# download the data

db_config = {'user': 'praktikum_student',         # user name
             'pwd': 'Sdf4$2;d-d30pp', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the data base

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

query_books = '''SELECT * FROM books'''
query_authors = '''SELECT * FROM authors'''
query_publishers = '''SELECT * FROM publishers'''
query_ratings = '''SELECT * FROM ratings'''
query_reviews = '''SELECT * FROM reviews'''

books = pd.io.sql.read_sql(query_books, con = engine)
authors = pd.io.sql.read_sql(query_authors, con = engine)
publishers = pd.io.sql.read_sql(query_publishers, con = engine)
ratings = pd.io.sql.read_sql(query_ratings, con = engine)
reviews = pd.io.sql.read_sql(query_reviews, con = engine)

# Study the tables

In [3]:
#printing the first 5 rows of books table
books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [4]:
#checking the portion of data on books table
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


In [5]:
#printing the first 5 rows of authors table
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [6]:
#checking the portion of data on authors table
authors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


In [7]:
#printing the first 5 rows of publishers table
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [8]:
#checking the portion of data on publishers table
publishers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


In [9]:
#printing the first 5 rows of ratings table
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [10]:
#checking the portion of data on ratings table
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


In [11]:
#printing the first 5 rows of reviews table
reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [12]:
#checking the portion of data on reviews table
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


# Task 1: Find the number of books released after January 1, 2000

In [13]:
#printing first 5 rows for my convenience
books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [14]:
#writing aqueery for the task
query_release_date = '''SELECT COUNT(book_id) 
                        FROM books
                        WHERE publication_date::date > '2000-01-01'; ''' 
pd.io.sql.read_sql(query_release_date, con = engine)

count
0    819

We can see that most of the books (819/1000) are released after 2000. 

# Task 2: Find the number of user reviews and the average rating for each book

In [15]:
#printing first 5 rows for my convenience
reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [16]:
#writing a query for the task
query_rating_review = '''SELECT 
books.book_id AS id,
books.title AS title,
AVG(ratings.rating) AS avg_rating,
COUNT(DISTINCT reviews.review_id) as review_cnt
FROM books
RIGHT JOIN  reviews ON books.book_id = reviews.book_id
RIGHT JOIN  ratings ON books.book_id = ratings.book_id
GROUP BY id, title
ORDER BY review_cnt DESC;
'''
pd.io.sql.read_sql(query_rating_review, con = engine)

id                                              title  avg_rating  \
0    948.0                            Twilight (Twilight  #1)    3.662500   
1    854.0                                           The Road    3.772727   
2    656.0                                     The Book Thief    4.264151   
3    734.0                                   The Glass Castle    4.206897   
4    963.0                                Water for Elephants    3.977273   
..     ...                                                ...         ...   
990  465.0                  Naked Empire (Sword of Truth  #8)    3.500000   
991  446.0                                Moo  Baa  La La La!    3.000000   
992  431.0                Merrick (The Vampire Chronicles #7)    4.000000   
993  196.0  Dragons of Winter Night (Dragonlance: Chronicl...    4.000000   
994    NaN                                               None    3.923077   

     review_cnt  
0             7  
1             6  
2             6  
3             6  
4             6  
..          ...  
990           1  
991           1  
992           1  
993           1  
994           0  

[995 rows x 4 columns]

We can see that the highest number of reviews for a book is 7. It seems readers dont tend to leave reviews so much. 

# Task 3: Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).

In [17]:
#writing aqueery for the task
query_publisher = '''SELECT
COUNT(books.book_id) as books_n,
publishers.publisher as publisher, publishers.publisher_id as id
FROM books
INNER JOIN publishers on books.publisher_id = publishers.publisher_id
WHERE num_pages > 50
GROUP BY publisher, id
ORDER BY books_n DESC
LIMIT 1;
'''
pd.io.sql.read_sql(query_publisher, con = engine)

books_n      publisher   id
0       42  Penguin Books  212

The publusher with the greatest number of books is 'Penguin Books', with 42 books published. 

# Task 4: Identify the author with the highest average book rating (look only at books with at least 50 ratings).

In [18]:
#writing aqueery for the task
query_author_rating = '''SELECT 
authors.author as author, 
books.author_id as id,
AVG(ratings.rating) as avg_rating
FROM books
RIGHT JOIN authors ON books.author_id = authors.author_id
RIGHT JOIN ratings ON ratings.book_id = books.book_id
GROUP BY author, id
HAVING COUNT(ratings.rating_id) > 50
ORDER BY avg_rating DESC
LIMIT 1;
'''
pd.io.sql.read_sql(query_author_rating, con = engine)

author   id  avg_rating
0  J.K. Rowling/Mary GrandPré  236    4.288462

The author with the highest average book rating is J.K. Rowling/Mary GrandPré. Also we can conclude that readers tend to give rating more then leave reviews for a book (the result contains al least 50 ratings while the maximum reviews for a book is 7). 

# Task 5:  Find the average number of text reviews among users who rated more than 50 books

In [19]:
#writing aqueery for the task
query_user_reviews = '''SELECT
AVG (sub.count_text) as avg_count_text
FROM (SELECT
ratings.username as name,
COUNT( reviews.text) as count_text
FROM ratings 
LEFT JOIN reviews ON ratings.book_id = reviews.book_id
AND ratings.username = reviews.username
GROUP BY name
HAVING COUNT(rating_id) >50) as sub
ORDER BY avg_count_text DESC
'''
pd.io.sql.read_sql(query_user_reviews, con = engine)

avg_count_text
0       24.333333

The average number of text reviews among users who rated more than 50 books is 24.3. I assume that people who tend to give more rating tend to leave more reviews. 

# conclusion and recommendations
I would recommend to choose books that were published in the last few years. It seems that people tend to give more rating then reviews. I would drecommend to develop the rating system and give it more emphasis then text reviews.   
Also I would recommend to interact more with the readers who tend to give more ratings and reviews then others. And simultaneously to encourage other readers to leave ratings. 
